### DISCUSSION TITLE
How do you use encoders with StreamFX (like ProRes)? Is it possible to use any encoder in OBS with StreamFX that FFmpeg supports, like x265?

### DISCUSSION THREAD

**Discussion by YousufSSyed** [Xaymar/obs-stream-effects#844] [General]
I spent a good amount of time trying to figure out how encoders work (or what they even do) with StreamFX. As far as I can tell, StreamFX lets you record with encoders it supports, correct? If so, I want to try using ProRes, I looked at two [wiki](https://github.com/Xaymar/obs-StreamFX/wiki/Encoder-FFmpeg-Apple-ProRes) [pages](https://github.com/Xaymar/obs-StreamFX/wiki/Encoder-FFmpeg), but they don't say what to do and are vague. 

While my end goal would be to record using x265 (H265 software encoding) on OBS, I think maybe figuring ProRes could help me do that, or at least see for myself how StreamFX encoders work. I also want to see if I could use any encoders that FFmpeg supports (whatever encoders are in said FFmpeg binary if there even is any, and if there is, does it support x265? If not, could I replace the FFmpeg binary).

**Comment by Xaymar** [comment]
The encoders StreamFX provides are all present in the encoder selection drop-down shown in Advanced Output Mode. The wiki pages explain the feature and their configuration - it is expected that the end-user already understands what OBS Studio does and how it works. See the endless amounts of tutorials available on YouTube if you do not already understand OBS Studio.

**Reply by Xaymar** [reply]
To answer the second question as well, it is possible to use any encoder FFmpeg offers with StreamFX as long as a proper handler is present. If no handler is present, the encoder will not be registered at all.

**Reply by YousufSSyed** [reply]
> The encoders StreamFX provides are all present in the encoder selection drop-down shown in Advanced Output Mode. The wiki pages explain the feature and their configuration - it is expected that the end-user already understands what OBS Studio does and how it works. 

Well, I already knew about advanced output mode but didn't know that the StreamFX encoders were in that same encoders menu*, because despite searching online trying to figure out where they are, I didn't find any answers (hence why I asked here). I found Prores in OBS with StreamFX installed, but when I uninstalled StreamFX (deleted the folder `/Library/Application Support/obs-studio/plugins/StreamFX` on Mac), I still saw the Prores encoder in the dropdown menu. 
* *Given those two things, can you understand why I didn't know that?

Can you explain what exactly StreamFX's encoder functionality does? Because I didn't find anything online on that either. 
* Well I did find the old [obs-ffmpeg-encoder](https://github.com/Xaymar/obs-ffmpeg-encoder) repo which says _"Adds ffmpeg encoders for normal use instead of requiring ffmpeg output."_, but the encoders specified there been added to the OBS encoders dropdown since.

> To answer the second question as well, it is possible to use any encoder FFmpeg offers with StreamFX as long as a proper handler is present. If no handler is present, the encoder will not be registered at all.

Can you explain this in more detail? Where would the handler be located? Could I add a handler myself to use whatever encoders I want, without OBS officially supporting them?

**Reply by Xaymar** [reply]
> Given those two things, can you understand why I didn't know that?

The encoders provided by StreamFX are only available when StreamFX is installed and loaded. If you had an encoder selected that is no longer present, OBS Studio may opt to still show it in order to not break their User Interface - but it will have no options visible. If the options are still visible, you didn't remove StreamFX entirely.

> Well I did find the old [obs-ffmpeg-encoder](https://github.com/Xaymar/obs-ffmpeg-encoder) repo

The obs-ffmpeg-encoders were merged into StreamFX. None of the encoders provided in either StreamFX or obs-ffmpeg-encoder are available in the Advanced Output Mode without using the FFmpeg Output Mode.

> Can you explain this in more detail? Where would the handler be located? Could I add a handler myself to use whatever encoders I want, without OBS officially supporting them?

Handlers are coded in C or C++, and are located here: https://github.com/Xaymar/obs-StreamFX/tree/master/source/encoders/handlers . As long as the FFmpeg binaries list the encoder, it is possible to provide a handler for the FFmpeg encoder integration, which will show up in the advanced output menu encoder selection (for recording).

**Reply by YousufSSyed** [reply]
Are there easier ways to add x265 to OBS for myself? With or without StreamFX? It looks like trying to add a handler would require some understanding of StreamFX's code (and OBS as well), and I think that could take hours at the very least.

**Reply by Xaymar** [reply]
That would be out-of-scope of this discussion forum. You may have better luck in the OBS Studio forums.

**Reply by YousufSSyed** [reply]
How about with StreamFX?

**Reply by Xaymar** [reply]
As explained above, you would need to either provide a handler, or write an integration for x265 directly.

**Reply by YousufSSyed** [reply]
OK, do u have documentation on how to write a handler and how they work? Or do u have any resources to point me to?

**Reply by Xaymar** [reply]
Writing a handler is as simple as implementing the interface. Take a look at the other existing handlers to see how you can write your own - most of it is just interfacing with FFmpeg and libOBS.

**Comment by YousufSSyed** [comment]
@Xaymar How do you uninstall StreamFX? because maybe I did and then reinstalled it, the encoder options under Custom Output (FFMpeg were the same). Also, do you know of a way to forcibly refresh the UI after uninstalling StreamFX?

Also, how does OBS use FFmpeg? Does it ship a binary with the app, if so where is it? If I were to write an x265 handler, would it just work? Because I imagine it wouldn't if the FFmpeg that OBS uses doesn't have x265 in it (for obvious reasons: patents and licensing).

**Reply by Xaymar** [reply]
You can uninstall StreamFX the same way you installed it. As Unix platforms don't usually have a central registry for installations not made via a locked-in system, you'll have to manually delete the StreamFX files from disk. The installer on MacOS places files into `/Library/Application Support/obs-studio/plugins`.

**Reply by Xaymar** [reply]
As for the other question(s), the FFmpeg binaries used by OBS Studio must expose the encoder you wish to use. Simple having a handler for an encoder alone does nothing, since there is no actual encoder to talk to. 

Note that StreamFX does not provide Custom Output (FFmpeg), and obs-ffmpeg-encoder also did not provide that tab. The encoders are simply listed in the normal encoder drop-down.

**Reply by YousufSSyed** [reply]
Where are the OBS FFmpeg binaries on Mac? I just looked it up but couldn't find anything. Not even the [Guide OBS Studio FFmpeg Upgrade](https://github.com/Xaymar/obs-StreamFX/wiki/Guide-OBS-Studio-FFmpeg-Upgrade) page says what to do on Mac.

**Reply by Xaymar** [reply]
As OBS Studio related things are out of scope for this forum, you may want to ask this question in either the off-topic section on our Discord, or on the actual OBS Studio discussion forums.

**Reply by YousufSSyed** [reply]
But I'll say, an x265 handler wouldn't work without changing the FFmpeg binary.

**Reply by Xaymar** [reply]
That is unfortunate, but not something StreamFX cares about. The guide you linked also was just a stop-gap to get Windows's FFmpeg up to speed with all other platforms (which were shipping 4.4 or even 5.0 already).
